plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBSnapshotTransaction.java

```diff
@@ -0,0 +1,155 @@
+import org.hyperledger.besu.plugin.services.metrics.OperationTimer;
+import org.hyperledger.besu.plugin.services.storage.KeyValueStorageTransaction;
+import org.hyperledger.besu.plugin.services.storage.rocksdb.RocksDBMetrics;
+import org.hyperledger.besu.plugin.services.storage.rocksdb.RocksDbKeyIterator;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.OptimisticTransactionDB;
+import org.rocksdb.ReadOptions;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+import org.rocksdb.Snapshot;
+import org.rocksdb.Transaction;
+import org.rocksdb.WriteOptions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RocksDBSnapshotTransaction implements KeyValueStorageTransaction, AutoCloseable {
+  private static final Logger LOG = LoggerFactory.getLogger(RocksDBSnapshotTransaction.class);
+  private static final String NO_SPACE_LEFT_ON_DEVICE = "No space left on device";
+
+  private final RocksDBMetrics metrics;
+  private final OptimisticTransactionDB db;
+  private final ColumnFamilyHandle columnFamilyHandle;
+  private final Transaction snapTx;
+  private final Snapshot snapshot;
+  private final WriteOptions writeOptions;
+  private final ReadOptions readOptions;
+
+  RocksDBSnapshotTransaction(
+      final OptimisticTransactionDB db,
+      final ColumnFamilyHandle columnFamilyHandle,
+      final RocksDBMetrics metrics) {
+    this.metrics = metrics;
+    this.db = db;
+    this.columnFamilyHandle = columnFamilyHandle;
+    this.snapshot = db.getSnapshot();
+    this.writeOptions = new WriteOptions();
+    this.snapTx = db.beginTransaction(writeOptions);
+    this.readOptions = new ReadOptions().setSnapshot(snapshot);
+  }
+
+  private RocksDBSnapshotTransaction(
+      final OptimisticTransactionDB db,
+      final ColumnFamilyHandle columnFamilyHandle,
+      final RocksDBMetrics metrics,
+      final Snapshot snapshot) {
+    this.metrics = metrics;
+    this.db = db;
+    this.columnFamilyHandle = columnFamilyHandle;
+    this.snapshot = snapshot;
+    this.writeOptions = new WriteOptions();
+    this.snapTx = db.beginTransaction(writeOptions);
+    this.readOptions = new ReadOptions().setSnapshot(snapshot);
+  }
+
+  public Optional<byte[]> get(final byte[] key) {
+    try (final OperationTimer.TimingContext ignored = metrics.getReadLatency().startTimer()) {
+      return Optional.ofNullable(snapTx.get(columnFamilyHandle, readOptions, key));
+    } catch (final RocksDBException e) {
+      throw new StorageException(e);
+    }
+  }
+
+  @Override
+  public void put(final byte[] key, final byte[] value) {
+    try (final OperationTimer.TimingContext ignored = metrics.getWriteLatency().startTimer()) {
+      snapTx.put(columnFamilyHandle, key, value);
+    } catch (final RocksDBException e) {
+      if (e.getMessage().contains(NO_SPACE_LEFT_ON_DEVICE)) {
+        LOG.error(e.getMessage());
+        System.exit(0);
```

Comment:
System.exit feels too brutal to me. Is there any chance this can corrupt db for instance in case an ongoing update is happening in parallel?  Maybe we don't access db from different threads.

Reply:
Good question.  This is the same behavior as RocksDBTransaction.  The right course of action isn't really clear since we do not have any disk space to work with.